libbeat/statestore/backend/memlog/diskstore.go

```diff
@@ -165,5 +290,408 @@ func (s *diskstore) LogOperation(op op) error {
+	}
+
+	trySyncPath(homePath)
+	return nil
+}
+
+// removeOldDataFiles sorts the data files by their update sequence number and
+// finally deletes all but the newest file from the storage directory.
+func (s *diskstore) removeOldDataFiles() {
+	for i := range s.oldDataFiles {
+		path := s.oldDataFiles[i].path
+		err := os.Remove(path)
+		if err != nil && !os.IsNotExist(err) {
+			s.log.With("file", path).Errorf("Failed to delete old data file: %v", err)
+			s.oldDataFiles = s.oldDataFiles[i:]
+			return
+		}
+	}
+	s.oldDataFiles = nil
+}
+
+// listDataFiles returns a sorted list of data files with txid per file.
+// The list is sorted by txid, in ascending order (taking integer overflows
+// into account).
+func listDataFiles(home string) ([]dataFileInfo, error) {
+	files, err := filepath.Glob(filepath.Join(home, "*.json"))
+	if err != nil {
+		return nil, err
+	}
+
+	var infos []dataFileInfo
+	for i := range files {
+		info, err := os.Lstat(files[i])
+		if err != nil {
+			return nil, err
+		}
+		if !info.Mode().IsRegular() {
+			continue
+		}
+
+		name := filepath.Base(files[i])
+		name = name[:len(name)-5] // remove '.json' extension
+
+		id, err := strconv.ParseUint(name, 10, 64)
+		if err == nil {
+			infos = append(infos, dataFileInfo{
+				path: files[i],
+				txid: id,
+			})
+		}
+	}
+
+	// empty or most recent snapshot was complete (old data file has been deleted)
+	if len(infos) <= 1 {
+		return infos, nil
+	}
+
+	// sort files by transaction ID
+	sortDataFileInfos(infos)
+	return infos, nil
+}
+
+// sortDataFileInfos sorts the slice by the files txid.
+func sortDataFileInfos(infos []dataFileInfo) {
+	sort.Slice(infos, func(i, j int) bool {
+		return isTxIDLessEqual(infos[i].txid, infos[j].txid)
+	})
+}
+
+// loadDataFile create a new hashtable with all key/value pairs found.
+func loadDataFile(path string, tbl map[string]entry) error {
+	if path == "" {
+		return nil
+	}
+
+	err := readDataFile(path, func(key string, state common.MapStr) {
+		tbl[key] = entry{value: state}
+	})
+	return err
+}
+
+func readDataFile(path string, fn func(string, common.MapStr)) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	var states []map[string]interface{}
+	dec := json.NewDecoder(f)
+	if err := dec.Decode(&states); err != nil {
+		return fmt.Errorf("corrupted data file: %v", err)
+	}
+
+	for _, state := range states {
+		keyRaw := state["_key"]
+		key, ok := keyRaw.(string)
+		if !ok {
+			continue
+		}
+
+		delete(state, "_key")
+		fn(key, common.MapStr(state))
+	}
+
+	return nil
+}
+
+// loadLogFile applies all recorded transaction to an already initialized
+// memStore.
+// The txid is the transaction ID of the last known valid data file.
+// Transactions older then txid will be ignored.
+// loadLogFile returns the last commited txid in logTxid and the total number
+// of operations in logCount.
+// An incomplete transaction is recorded at the end of the log file, if
+// complete is false.
```

Comment:
I'm not sure what "complete is false" is referring to here, is this an outdated comment or is it referring to a value this PR doesn't reference explicitly?

Reply:
this can be removed. It's an artifact of an older change